internal/chezmoi/findexecutable.go

```diff
@@ -0,0 +1,47 @@
+package chezmoi
+
+import (
+	"os"
+	"path/filepath"
+	"sync"
+)
+
+var (
+	foundExecutableCacheMutex sync.Mutex
+	foundExecutableCache      = make(map[string]struct{})
```

Comment:
Yeah this was basically trying to follow the convention of `lookPath` so I just adapted it. I did think about it. I didn't want to hash the sequence of paths provided so it only makes sense for reducing `stat` calls but realistically it can be completely removed.

However I agree, if kept, it should be moved up to before the 2nd loop and made into `map[string]string`. The problem being I wasn't consistently thorough when converting this PR. Sorry. Will address soon. However; should we even bother cache?

Reply:
> I believe the caching here makes the results more nondeterministic and, depending on the path list and filesystem, will not reduce Stat calls.

Okay I reviewed this. I'm not sure I understand the case where it becomes more nondeterministic. -- It does assume a static file system during it's execution. To which the only solution I can think of is the removal of the cache entirely?

Since most of the review is around the cache I can't really proceed with this. I'm looking for direction here. @twpayne @halostatue . I am considering removing the cache as I am not sure if provides value.

Reply:
I wouldn't remove the cache, I would make it stricter. That is, for a given `(needle, haystack[])` tuple,  there is one guaranteed result for one run of `chezmoi apply`. In general, it means that your key would be something like `key := needle + "\x01" + strings.Join(haystack, "\x00")`. In this way, if I’m doing:

```
{{ findExecutable "vim" (list "/opt/local/bin" "/opt/homebrew/bin" "/usr/local/bin" "/usr/bin") }}
```

the result will be the same for every run of that parameter combination if any one of those returns a value (say it's `/opt/homebrew/bin/vim`). It's been a while, but I believe that right now, you’re still searching all of the parameters.

Reply:
Essentially, the improved caching will come from the approach I suggested in [an earlier comment](https://github.com/twpayne/chezmoi/pull/3162#discussion_r1287848282).